core/message.go

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package rocketmq
+
+//#cgo LDFLAGS: -L/usr/local/lib/ -lrocketmq
+//#include "rocketmq/CMessage.h"
+//#include "rocketmq/CMessageExt.h"
+import "C"
+import "fmt"
+
+type Message struct {
+	Topic string
+	Keys  string
+	// improve: maybe []byte is better.
+	Body string
+}
+
+func (msg *Message) String() string {
+	return fmt.Sprintf("[topic: %s, keys: %s, body: %s]", msg.Topic, msg.Keys, msg.Body)
+}
+
+type MessageExt struct {
+	Message
+	MessageID string
+	Tags      string
+	// improve: is there is a method convert c++ map to go variable?
+	cmsgExt *C.struct_CMessageExt
+	//Properties  string
+}
+
+func (msgExt *MessageExt) String() string {
+	return fmt.Sprintf("[messageId: %s, %s, Tags: %s]", msgExt.MessageID, msgExt.Message, msgExt.Tags)
+}
+
+func (msgExt *MessageExt) GetProperty(key string) string {
+	return C.GoString(C.GetMessageProperty(msgExt.cmsgExt, C.CString(key)))
+}
+
+func cmsgToGo(cmsg *C.struct_CMessage) *Message {
+	defer C.DestroyMessage(cmsg)
+	gomsg := &Message{}
```

Comment:
Why cmsg is only been destory and nothing to do with gomsg?

Reply:
this covert method will be implement asap.

Reply:
IMO，the go users only need to use go-variable, C.struct_CMessage will not use after converting, so we should free memory of c allocated manually at once.